packages/ddp-server/livedata_server.js

```diff
@@ -1076,16 +1078,20 @@ Object.assign(Subscription.prototype, {
     }
 
     // Did the handler call this.error or this.stop?
-    if (self._isDeactivated())
-      return;
-
-    //Both conventional and async publish handler functions are supported.
-    //If an object is returned with a then() function, it is either a promise or thenable
-    //and will be resolved asynchronously.
-    if (resMaybePromise && typeof resMaybePromise.then === "function") {
-      Promise.resolve(resMaybePromise).then(self._publishHandlerResult, e => self.error(e));
+    if (self._isDeactivated()) return;
+
+    // Both conventional and async publish handler functions are supported.
+    // If an object is returned with a then() function, it is either a promise
+    // or thenable and will be resolved asynchronously.
+    const isThenable =
+      resultOrThenable && typeof resultOrThenable.then === 'function';
+    if (isThenable) {
+      Promise.resolve(resultOrThenable).then(
+        (...args) => self._publishHandlerResult.bind(self)(...args),
```

Comment:
@vlasky this is the fix.

without this change it was not working for async functions, the this was Meteor instead of the Subscription.

Reply:
@filipenevola belated thanks for sorting that out. For some reason, when I tested with our MySQL-based publications, they  worked in async mode without the bind().